compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -331,15 +336,113 @@ struct GPUSetEncodingOpLoweringConversion
     // We want to make the shape consistent, so we need to append it with a
     // `collapse_shape` and a `expand_shape`, just to be conformant with how we
     // materialize for Flow and HAL op.
-    tensor::ExpandShapeOp result =
-        reshapeForInnerDims(rewriter, transposeOp->getResult(0),
-                            packOp->getDestType(), packOp->getSourceRank());
+    tensor::ExpandShapeOp result = reshapeForInnerDims(
+        rewriter, transposeOp->getResult(0),
+        cast<RankedTensorType>(transposeOp->getResult(0).getType()),
+        packOp->getDestType(), packOp->getSourceRank());
 
     rewriter.replaceOp(encodingOp, result);
     return success();
   }
 };
 
+struct GPUUnsetEncodingOpLoweringConversion
+    : public OpMaterializeEncodingPattern<IREE::Encoding::UnsetEncodingOp> {
+  using OpMaterializeEncodingPattern<
+      IREE::Encoding::UnsetEncodingOp>::OpMaterializeEncodingPattern;
+
+  LogicalResult
+  matchAndRewrite(IREE::Encoding::UnsetEncodingOp unsetEncodingOp,
+                  OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto converter = static_cast<const MaterializeEncodingTypeConverter *>(
+        getTypeConverter());
+
+    Location loc = unsetEncodingOp.getLoc();
+
+    FailureOr<MaterializeEncodingInfo> maybeEncodingInfo =
+        converter->getEncodingInfo(unsetEncodingOp.getSource().getType());
+    if (failed(maybeEncodingInfo)) {
+      return rewriter.notifyMatchFailure(unsetEncodingOp,
+                                         "unhandled result encoding");
+    }
+    SmallVector<int64_t> innerTiles = maybeEncodingInfo->innerTileSizes;
+    SmallVector<int64_t> intrinsicVectorShape =
+        maybeEncodingInfo->innerTileShapes;
+
+    // compute sourceShape:
+    auto srcConvertedType =
+        cast<RankedTensorType>(adaptor.getSource().getType());
+    SmallVector<int64_t> unpackSrcShape(
+        srcConvertedType.getShape().take_front(maybeEncodingInfo->srcRank));
+    unpackSrcShape.append(maybeEncodingInfo->innerTileSizes.begin(),
+                       maybeEncodingInfo->innerTileSizes.end());
+    auto unpackSrcType = RankedTensorType::get(
+        unpackSrcShape, unsetEncodingOp.getSourceType().getElementType());
+
+    auto iT1 = intrinsicVectorShape[0];
+    auto iT2 = intrinsicVectorShape[1];
+    auto oT1 = unpackSrcShape[2] / iT1;
+    auto oT2 = unpackSrcShape[3] / iT2;
+    SmallVector<int64_t> transposeSourceDims = {
+        unpackSrcShape[0], unpackSrcShape[1], oT1, iT1, oT2, iT2};
+    assert(transposeSourceDims.size() == 6);
+
+    size_t targetRank = unsetEncodingOp.getResultType().getRank();
+
+    SmallVector<int64_t> transposePerm;
+    transposePerm.push_back(0);
+    transposePerm.push_back(1);
```

Comment:
It's fine for now, but note that we need to kill these magic numbers at some moment.

Reply:
I will get a refactor patch to do these